pkg/brbct/brbct.go

```diff
@@ -0,0 +1,341 @@
+		}
+		initialize(&state, id)
+
+		fromId, err := strconv.ParseInt(from.Pb(), 10, 64)
+		if err != nil {
+			return err
+		}
+
+		dsl.MerkleProofVerifyRequest(m, mc.MerkleProofVerifier, rootHash, chunk, proof, &hashEchoMessageContext{id: id, fromId: fromId, chunk: chunk, rootHash: rootHash, proof: proof})
+		return nil
+	})
+
+	dsl.MerkleProofVerifyResult(m, func(result bool, context *hashEchoMessageContext) error {
+		if context.id <= lastId {
+			return nil
+		}
+		currentState := state[context.id]
+		if result {
+			updateAccumulator(
+				&currentState.echos,
+				&currentState.echoMessagesCount,
+				&currentState.echoMessagesAccumulator,
+				string(context.rootHash),
+				context.chunk,
+				context.fromId,
+				params,
+			)
+			return nil
+		} else {
+			return fmt.Errorf("received invalid echo")
+		}
+	})
+
+	brbpbdsl.UponReadyMessageReceived(m, func(from t.NodeID, id int64, rootHash []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		initialize(&state, id)
+
+		fromId, err := strconv.ParseInt(from.Pb(), 10, 64)
+		if err != nil {
+			return err
+		}
+
+		stringRepl := string(rootHash)
+
+		currentState := state[id]
+
+		if _, ok := currentState.readyMessagesReceived[stringRepl]; !ok {
+			currentState.readyMessagesReceived[stringRepl] = make([]bool, params.GetN())
+		}
+		if !currentState.readyMessagesReceived[stringRepl][fromId] {
+			currentState.readyMessagesReceived[stringRepl][fromId] = true
+
+			currentState.readyMessagesCount[stringRepl]++
+			if currentState.readyMessagesCount[stringRepl] > currentState.readyMessagesAccumulator.count {
+				currentState.readyMessagesAccumulator.count = currentState.readyMessagesCount[stringRepl]
+				currentState.readyMessagesAccumulator.value = stringRepl
+			}
+		}
+
+		return nil
+	})
+
+	dsl.UponCondition(m, func() error {
+		for id, currentState := range state {
+			if id <= lastId {
+				continue
+			}
+			if (currentState.echoMessagesAccumulator.count >= params.GetN()-params.GetF()) && !currentState.sentReady {
+				currentState.sentReady = true
+				eventpbdsl.SendMessage(m, mc.Net, brbmsgs.ReadyMessage(mc.Self, id, []byte(currentState.echoMessagesAccumulator.value)), params.AllNodes)
+			}
+
+			if (currentState.readyMessagesAccumulator.count >= params.GetN()-2*params.GetF()) && !currentState.sentReady {
+				currentState.sentReady = true
+				eventpbdsl.SendMessage(m, mc.Net, brbmsgs.ReadyMessage(mc.Self, id, []byte(currentState.readyMessagesAccumulator.value)), params.AllNodes)
+			}
+
+			if currentState.readyMessagesAccumulator.count >= params.GetN()-params.GetF() && currentState.echoMessagesAccumulator.count > params.GetN()-2*params.GetF() && !currentState.delivered {
+				currentState.delivered = true
+				decoded := make([][]byte, params.GetN())
+				for i := 0; i < params.GetN(); i++ {
+					decoded[i] = currentState.echos[currentState.echoMessagesAccumulator.value][i]
+				}
+
+				err := encoder.Reconstruct(decoded)
+				if err != nil {
+					return err
+				}
```

Comment:
I think there is a missing check before delivery: you need to recompute the Merkle tree root from the decoded data and verify it against the expected Merkle tree root. Otherwise, the protocol would be vulnerable to a malicious leader not performing the erasure-correcting encoding correctly.